### PR TITLE
[MXNET-490] Added OpenLSTMRNN together with benchmarks and Tensorboard callback routines.

### PR DESCRIPTION
**Pull request by ArmageddonKnight**
## Description ##
This PR added an open-source LSTM RNN implementation, and also included benchmarks for comparison between different LSTM RNN backends and tools for measurements. This [**arxiv paper**](https://arxiv.org/pdf/1805.08899.pdf) contains more implementation details. 

The end-to-end training throughput comparison on the **Wikitext-2 language-modeling benchmark** between MXNet default implementation (**Default**), cuDNN 6 (**cuDNN**) and OpenLSTMRNN (**Open**) implementation of LSTM RNN is shown in the diagram below:

<p>
  
</p>

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Added `OpenLSTMRNN`, an LSTM RNN CUDA implementation to the MXNet C++ core library (`./src/operator/contrib/`).
- [x] Added `TensorboardSpeedometer` and `TensorboardLogValidationMetricsCallback` for logging training and validation scores as training progresses.
- [x] Added benchmarks in `example` directory for comparison between MXNet default, cuDNN, and `OpenLSTMRNN` backend implementations. 

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: example/rnn-backends/bucketing/dataset/download_ptb.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# This file downloads the PTB dataset.
+
+mkdir -p $(cd $(dirname $0) && pwd)/ptb
+
+cd $(cd $(dirname $0) && pwd)/ptb && \
```
Comment (szha) [inline example/rnn-backends/bucketing/dataset/download_ptb.sh]: we're moving away from ptb. see https://github.com/apache/incubator-mxnet/pull/11435

**Comment by lupesko** [comment]
@ArmageddonKnight thanks for your contribution!
Looks like some CI tests failed, can you look into it so we can move this forward?

**Comment by ArmageddonKnight** [comment]
@lupesko Thank you so much for your comment. I have revised accordingly and passed the CI tests.

**Comment by lupesko** [comment]
Thank you @ArmageddonKnight !
Bouncing to get reviews: @szha @sandeep-krishnamurthy @anirudh2290

**Comment by vandanavk** [comment]
@ThomasDelteil @safrooze for review

**Comment by safrooze** [comment]
@ArmageddonKnight Can you comment on what the benefit of this implementation is? It would be great if the Readme on example would include a comparison table for a few LSTM configurations. It's not really clear why one would use OpenLSTM. Also, is this implementation based on a paper or some reference?

**Comment by ArmageddonKnight** [comment]
@safrooze Thank you for your reply. The reason is because in our experiments we see that OpenLSTM is able to beat cuDNN 6 in end-to-end performance evaluation (in terms of training throughput, the benchmarks used are included in this commit). Besides, we also think that it is good to have an open-source LSTM CUDA implementation that other researchers can experiment with or use as a baseline.

The reference is [**here**](https://arxiv.org/pdf/1805.08899.pdf).

**Comment by safrooze** [comment]
@ArmageddonKnight This is great! Can you please add the link to arxiv to the PR description and also include a comparison table in the description if you have already done a benchmark?

**Comment by ArmageddonKnight** [comment]
@safrooze Thanks a lot for your suggestions. I have made changes to the descriptions accordingly.

**Comment by vandanavk** [comment]
@ArmageddonKnight 
Please execute pylint on these files using ci/other/pylintrc in incubator-mxnet folder and fix errors (if any).
Example:
`pylint --rcfile=ci/other/pylintrc --ignore-patterns="..so$$,..dll$$,..dylib$$" example/rnn_backends/word_lm/*.py`

**Comment by ArmageddonKnight** [comment]
@vandanavk Hi, I have run the pylint checking on all the submitted python files and fixed the warnings accordingly.

```
example/rnn-backends/word_lm$ pylint --rcfile=../../../ci/other/pylintrc --ignore-patterns="..so$$,..dll$$,..dylib$$" *.py

-------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

```
example/rnn-backends/bucketing$ pylint --rcfile=../../../ci/other/pylintrc --ignore-patterns="..so$$,..dll$$,..dylib$$" *.py

-------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

```
python/mxnet/rnn$ pylint --rcfile=../../../ci/other/pylintrc --ignore-patterns="..so$$,..dll$$,..dylib$$" rnn_cell.py

------------------------------------
Your code has been rated at 10.00/10
```

```
python/mxnet$ pylint --rcfile=../../ci/other/pylintrc --ignore-patterns="..so$$,..dll$$,..dylib$$" callback.py

------------------------------------
Your code has been rated at 10.00/10
```

```
python/mxnet$ pylint --rcfile=../../ci/other/pylintrc --ignore-patterns="..so$$,..dll$$,..dylib$$" initializer.py

------------------------------------
Your code has been rated at 10.00/10
```

File: src/operator/contrib/cu_open_lstm_rnn-inl.cuh [hunk 2]
```diff
@@ -0,0 +1,1125 @@
...
+                   RE_CAST(m_i2h_grad_workspace.dptr) +
+                       seq_idx * m_num_hidden_units_x_batch_size,
+                   param_.state_outputs ?
+                     RE_CAST(m_h2h_grad_workspace.dptr) +
+                       layer_idx * m_num_hidden_units_x_batch_size :
+                       NULL,
+                   param_.batch_size,
+                   m_algo);
+          } else {
+            /*
```
Comment (apeforest) [inline src/operator/contrib/cu_open_lstm_rnn-inl.cuh]: Remove commented code

File: src/operator/contrib/cu_open_lstm_rnn-inl.cuh [hunk 3]
```diff
@@ -0,0 +1,1125 @@
...
+                       seq_idx * m_num_hidden_units_x_batch_size,
+                   RE_CAST(m_h2h_grad_workspace.dptr) +
+                     layer_idx * m_num_hidden_units_x_batch_size,
+                   param_.batch_size,
+                   m_algo);
+          }  // if (seq_idx == static_cast < int > (param_.seq_len - 1))
+          // record that we are computing m_i2h_workspace
+          CUDA_CALL(cudaEventRecord(m_event_h2h[layer_idx][seq_idx],
+                                    m_stream_h2h[layer_idx]));
+          /*
```
Comment (apeforest) [inline src/operator/contrib/cu_open_lstm_rnn-inl.cuh]: cleanup commented code

File: src/operator/contrib/cu_open_lstm_rnn-inl.cuh [hunk 4]
```diff
@@ -0,0 +1,1125 @@
...
+                      layer_idx * 4 * m_num_hidden_units_x_batch_size,
+                    RE_CAST(m_hidden.dptr) +
+                      layer_idx * (param_.seq_len + 1) * m_num_hidden_units_x_batch_size +
+                                          seq_idx      * m_num_hidden_units_x_batch_size,
+                    CUBLAS_OP_N, CUBLAS_OP_T,
+                    4 * param_.num_hidden_units, param_.batch_size,
+                        param_.num_hidden_units, param_.batch_size,
+                    1.0, 1.0);
+          if (seq_idx > 0) {
+            /*
```
Comment (apeforest) [inline src/operator/contrib/cu_open_lstm_rnn-inl.cuh]: cleanup

File: src/operator/contrib/cu_open_lstm_rnn-inl.cuh [hunk 5]
```diff
@@ -0,0 +1,1125 @@
...
+        // Input -> Hidden
+        //=====================================================================
+        CUBLAS_CALL(cublasSetStream(m_cublas_handle, m_stream_i2h[layer_idx]));
+        // wait here until the data in m_i2h_workspace is ready
+        // i2h needs to wait for h2h
+        for (int seq_idx = seq_rbegin; seq_idx > seq_rend; --seq_idx)
+          CUDA_CALL(cudaStreamWaitEvent(m_stream_i2h[layer_idx],
+                                        m_event_h2h[layer_idx][seq_idx], 0));
+        if (layer_idx > 0) {
+          /*
```
Comment (apeforest) [inline src/operator/contrib/cu_open_lstm_rnn-inl.cuh]: cleanup

File: src/operator/contrib/cu_open_lstm_rnn-inl.cuh [hunk 6]
```diff
@@ -0,0 +1,1125 @@
...
+              &m_i2h_grad_workspace[seq_rend + 1],
+              &m_i2h_grad_workspace[seq_rend + 1],
+              &m_i_dp_uniform_rv[layer_idx - 1][seq_rend + 1],
+              m_stream_i2h[layer_idx], seq_rbegin - seq_rend);
+          }
+           */
+          for (int seq_idx = seq_rbegin; seq_idx > seq_rend; --seq_idx)
+            matmul_ex(m_cublas_handle,
+                      i2h_weight_grad.dptr_ +
+                        4 * param_.num_hidden_units * param_.embed_dim +
```
Comment (apeforest) [inline src/operator/contrib/cu_open_lstm_rnn-inl.cuh]: nit: align line with previous

File: src/operator/contrib/cu_open_lstm_rnn-inl.cuh [hunk 7]
```diff
@@ -0,0 +1,1125 @@
...
+                  (layer_idx - 1) * param_.seq_len * m_num_hidden_units_x_batch_size +
+                                    (seq_rend + 1) * m_num_hidden_units_x_batch_size,
+                param_.i_dp_prob, m_num_hidden_units_x_batch_size * (seq_rbegin - seq_rend));
+          }  // if (param_.i_dp_prob != 0)
+          // record that we are computing m_i2h_grad_workspace
+          for (int seq_idx = seq_rbegin; seq_idx > seq_rend; --seq_idx)
+            CUDA_CALL(cudaEventRecord(m_event_i2h[layer_idx][seq_idx],
+                                      m_stream_i2h[layer_idx]));
+        } else {
+          /*
```
Comment (apeforest) [inline src/operator/contrib/cu_open_lstm_rnn-inl.cuh]: cleanup

File: src/operator/contrib/cu_open_lstm_rnn-inl.cuh [hunk 8]
```diff
@@ -0,0 +1,1125 @@
...
+                                   1.0, 0.0);
+        }  // if (layer_idx > 0)
+      }  // layer_idx
+    }  // schedule
+    CUDA_CALL(cudaMemcpyAsync(i2h_bias_grad.dptr_,
+                              h2h_bias_grad.dptr_,
+                              param_.num_layers * 4 * param_.num_hidden_units * sizeof(float),
+                              cudaMemcpyDeviceToDevice,
+                              m_stream_h2h[0]));
+    /*
```
Comment (apeforest) [inline src/operator/contrib/cu_open_lstm_rnn-inl.cuh]: clean up

File: src/operator/contrib/cu_open_lstm_rnn-inl.cuh [hunk 9]
```diff
@@ -0,0 +1,1125 @@
...
+      //=======================================================================
+      // infer the parameters from the input data
+      param_.batch_size = data.shape_[0];
+      param_.seq_len    = data.shape_[1];
+      param_.embed_dim  = data.shape_[2];
+      m_num_hidden_units_x_batch_size = param_.batch_size * param_.num_hidden_units;
+      //=======================================================================
+      // allocate workspace
+      m_data_T_major       = Storage::Get()->Alloc(param_.seq_len * param_.embed_dim *
+                                                     param_.batch_size * sizeof(float),
```
Comment (apeforest) [inline src/operator/contrib/cu_open_lstm_rnn-inl.cuh]: nit: align

File: src/operator/contrib/open_lstm_rnn_include/lstm_cell.cuh [hunk 10]
```diff
@@ -0,0 +1,499 @@
...
+    //   forget_gate * prev_cell_state[g_threadIdx] +
+    //   input_gate * input_actv;
+    float forget_gate_grad = curr_cell_state_grad *
+                             prev_cell_state[g_threadIdx];
+    float  input_gate_grad = curr_cell_state_grad * input_actv;
+    float  input_actv_grad = curr_cell_state_grad * input_gate;
+
+    cell_grad[g_threadIdx] = curr_cell_state_grad * forget_gate;
+
+    /*
```
Comment (apeforest) [inline src/operator/contrib/open_lstm_rnn_include/lstm_cell.cuh]: clean up

File: src/operator/contrib/open_lstm_rnn_include/lstm_cell.cuh [hunk 11]
```diff
@@ -0,0 +1,499 @@
...
+    float forget_gate = __cu_sigmoidf(gate_input[1]);
+    float  input_actv =          tanh(gate_input[2]);
+    float output_gate = __cu_sigmoidf(gate_input[3]);
+     */
+     input_gate_input_grad =  input_gate_grad *  input_gate * (1 -  input_gate);
+    forget_gate_input_grad = forget_gate_grad * forget_gate * (1 - forget_gate);
+     input_actv_input_grad =  input_actv_grad * (1 - input_actv * input_actv);
+    output_gate_input_grad = output_gate_grad * output_gate * (1 - output_gate);
+
+    /*
```
Comment (apeforest) [inline src/operator/contrib/open_lstm_rnn_include/lstm_cell.cuh]: clean up

File: src/operator/contrib/open_lstm_rnn.cc [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+namespace mxnet {
+namespace op {
+
+template <>
+Operator *CreateOp<cpu>(OpenLSTMRNNParam param, int dtype) {
+  LOG(FATAL) << "OpenLSTMRNN is only available for gpu at the moment.";
+
+  Operator * op = nullptr;
+
+  MSHADOW_REAL_TYPE_SWITCH(dtype, DType,
```
Comment (apeforest) [inline src/operator/contrib/open_lstm_rnn.cc]: why is this needed?

File: src/operator/contrib/open_lstm_rnn.cc [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+ * \author Bojian (Jack) Zheng, Gennady Pekhimenko
+ */
+#include "./open_lstm_rnn-inl.h"
+
+namespace mxnet {
+namespace op {
+
+template <>
+Operator *CreateOp<cpu>(OpenLSTMRNNParam param, int dtype) {
+  LOG(FATAL) << "OpenLSTMRNN is only available for gpu at the moment.";
```
Comment (apeforest) [inline src/operator/contrib/open_lstm_rnn.cc]: change to LOG(ERROR)

File: src/operator/contrib/open_lstm_rnn-inl.h [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+    return {};
+  }
+
+  std::vector<ResourceRequest> BackwardResource(
+    const std::vector<TShape> &in_shape) const override {
+    return {};
+  }
+
+  Operator * CreateOperator(Context ctx) const override {
+    LOG(FATAL) << "OpenLSTMRNN can only run on the GPU.";
```
Comment (apeforest) [inline src/operator/contrib/open_lstm_rnn-inl.h]: use LOG(ERROR)

File: src/operator/contrib/open_lstm_rnn.cc (hunk 12 above)
Comment (ArmageddonKnight) [reply src/operator/contrib/open_lstm_rnn.cc]: @apeforest This was written analogously with the [`cuDNN` implementation of LSTM RNN](https://github.com/apache/incubator-mxnet/blob/v0.12.0/src/operator/rnn.cc).

Comment (ArmageddonKnight) [reply src/operator/contrib/open_lstm_rnn.cc]: @apeforest This was written analogously with the [`cuDNN` implementation of LSTM RNN](https://github.com/apache/incubator-mxnet/blob/v0.12.0/src/operator/rnn.cc).

File: src/operator/contrib/cu_open_lstm_rnn-inl.cuh (hunk 9 above)
Comment (ArmageddonKnight) [reply src/operator/contrib/cu_open_lstm_rnn-inl.cuh]: @apeforest I deliterately indented those lines. The reason is because those are continuations of current parameter, and the following parameters are all perfectly aligned. This can help make the code more readable.

File: src/operator/contrib/cu_open_lstm_rnn-inl.cuh [hunk 15]
```diff
@@ -0,0 +1,907 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*!
+ * Copyright (c) 2018 by Contributors
+ * \file open_lstm_rnn-inl.cuh
+ * \brief LSTM RNN Open-Source CUDA Implementation
+ * \author Bojian (Jack) Zheng, Gennady Pekhimenko, Jeremy Appleyard
+ */
+#ifndef MXNET_OPERATOR_CONTRIB_CU_OPEN_LSTM_RNN_INL_CUH_
```
Comment (apeforest) [inline src/operator/contrib/cu_open_lstm_rnn-inl.cuh]: I think a convention of the operator header file name does not start with cu_...

File: example/rnn-backends/word_lm/model.py [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+    # The implementation also messed up with the order of cell and hidden state,
+    # because according to the state order specified in rnn_cell.py (+L682),
+    # cell state should appear AFTER hidden state in FusedRNNCell. However, this has zero effect
+    # on the final training results because they are both zero-initialized.
+
+    states = []
+    state_names = []
+
+    # for i in range(num_layers):
+    #     prefix = 'lstm_l%d_' % i
```
Comment (Roshrini) [inline example/rnn-backends/word_lm/model.py]: Remove commented block?

**Comment by Roshrini** [comment]
@szha @eric-haibin-lin @apeforest Can you take a look once if its ready to merge?

File: example/rnn-backends/word_lm/model.py (hunk 16 above)
Comment (ArmageddonKnight) [reply example/rnn-backends/word_lm/model.py]: @Roshrini I am terribly sorry for the late response. I do not this it is appropriate for me to remove this part of the comments because they are used to show the discrepencies between the original word language-modeling benchmark and the benchmark that I am using (i.e. it is deliberately left there for **comparison** purpose).

Comment (ArmageddonKnight) [reply example/rnn-backends/word_lm/model.py]: Previously when working on benchmarking, I noticed that there are inefficiencies in the original implementation of the word-level language modeling benchmark (as was discussed here: https://github.com/apache/incubator-mxnet/issues/10304#issuecomment-377031511). I therefore came up with my own implementation that achieves the same validation score on the same dataset. The commented code-block is used to show **how the original implementation and my implementation do the same thing algorithmically**.

**Comment by ArmageddonKnight** [comment]
@ankkhedia Thank you so much for your reminder. I have addressed her feedback.

**Comment by ankkhedia** [comment]
@Roshrini It seems that your query has been addressed

@safrooze @apeforest @szha Could you please look into the PR if your concerns has been addressed.

**Comment by anirudhacharya** [comment]
@apeforest @szha pinging for another round of review

**Comment by stu1130** [comment]
@ThomasDelteil @thomelane could you review please?

**Comment by vandanavk** [comment]
@roywei for review

**Comment by sandeep-krishnamurthy** [comment]
@szha @eric-haibin-lin - Can you please help review this PR?

File: python/mxnet/initializer.py [hunk 17]
```diff
@@ -161,7 +161,7 @@ def _legacy_init(self, name, arr):
         Parameters
         ----------
         name : str
-            Name of corresponding NDArray.
+            Name of corrosponding NDArray.
```
[the remark this answers is not in the document]
Comment (safrooze) [reply python/mxnet/initializer.py]: Minor spelling correction in docstring, otherwise LGTM.

**Comment by safrooze** [comment]
@ArmageddonKnight are you planning to re-submit this PR?

**Comment by ArmageddonKnight** [comment]
@safrooze Thanks for your response. I am currently making changes to the source code and therefore hoping to temporarily close this pull request.
